### ISSUE TITLE
Option to Disable commit in execute()

### ISSUE THREAD

**Issue by ludz-lim** [ThibTrip/pangres#44]
It would be great if there's a way to disable `commit()` when the `execute()` is called (See line 162 of upsert.py).

This might be useful if the `upsert()` is part of a transaction

**Comment by ThibTrip** [comment]
Hi @ludz-lim ! Thanks, that is a good suggestion 👍!

I think it is doable but I need to test it because I see some challenges:
1. `connection.commit` does not really exist yet (it's for sqlalchemy 2.0 or when using the future parameter when creating an engine or with asynchronous engines). I suppose I can use a transaction and give back a connection from the engine though
2. one could imagine giving back the connection at each chunk upserted without committing (`pangres` works with chunks by default) but we already have this parameter `yield_chunks` that gives the proxies objects back (results of executed queries) for accessing for instance the number of affected rows. So here I would only give back the connection after everything has been "executed" but not committed yet.
3. what happens i we insert millions of rows without committing (because there were many chunks)? Is something going to break at some point e.g. due to the max number of sql parameters?

**Comment by ThibTrip** [comment]
I have some updates @ludz-lim. I've numbered my points so I can give some updates on them:

1. Somehow despite all my attempts the change always get committed. Not sure what I am doing wrong and I am pretty sure this is possible as I have done commit-as-you-go workflows in the past. I wonder if it is because it is inside of a function (I am of course not using a context manager). I have tried `connection.begin`, the `future=True` flag on the engine, using `Session` and `engine.connect().execution_options(autocommit=False)`. I'll try all variants again in a dedicated notebook and check if it works at least on the main scope.
2. So if we have a parameter with `commit=True` and we also have `yield_chunks=True` then an error should be raised
3. I don't know about the millions of rows yet but at least I can make a custom exception and embed the `connection` inside of the exception so that the user can at least `rollback` and `close` in case something goes wrong.

**Comment by ThibTrip** [comment]
Hi @ludz-lim, sorry I realized I misunderstood this issue! What I had in mind was a multiple execution scenario (which would be something else entirely where one would even have to work at cursor level).

I think what you want to achieve is a scenario similar to the code below where the `upsert` operation from `pangres` gets rollbacked in case some exception occurs, right?

```python
with engine.begin() as trans:
    upsert(...)
    trans.commit()
    # do some operation that relies on the upsert
   ...
    # execute something else
    trans.execute('INSERT INTO...')
    trans.commit()
```

First i would need to do deep changes in the library so that everything receives a `connection` instead of an `engine`. Turns out it is something that I should do anyways because:

1. I noticed I am creating connections all over the place in the library which is probably inefficent
2. If I want to go async then for some methods such as altering data types of empty columns then I cannot have a statement like this (it is not async compatible):

```python
with engine.connect() as connection:
    ...
```

But this will take a long time and I am very busy at the moment even when not working :turtle:.

**Comment by ThibTrip** [comment]
Hi @ludz-lim, I have published a new version of pangres (v3.1) where all operations run within a transaction and using a single connection.
The next step is to change the first argument of the `upsert` function to a generic `con` instead of `engine`. This indicates that one can also use a `Connection` object allowing for more control over connections and transactions.

Since this is a breaking change I will first introduce a `upsert_future` function in the next version so that people can work with it (I still need to do a few tests before I can publish it). Then a while later I'll publish a new major version which replaces `upsert`.

See this [demo notebook](https://github.com/ThibTrip/pangres/blob/transaction_and_connection_control/demos/transaction_control.ipynb).
I'd be glad to have your opinion.

**Comment by ThibTrip** [comment]
I realized that this sort of deprecation cycle I had in mind did not make much sense. You can now install version 4.0 and it should work exactly as you wanted it :rocket:. Feel free to reopen this if it is not the case.
